# Data Analysis and Model Building

### Vishal Kumar : kumar.vishal318@gmail.com
### Omkarnath Parida : pomkarnath98@gmail.com

In [1]:
import pandas as pd
import numpy as np

In [2]:
#reading dataset
insta = pd.read_csv("./clean_data.csv")
insta.head()

username  mutual_connecting  total_following  total_followers  \
0   7olga7777777                  0              107               69   
1    aadamkathst                  2              212               40   
2  aadish___khan                  2              116               65   
3     aarzoo6466                  2              222               40   
4   abbasyadavad                  0               25               17   

              name  private_account  business_account  recently_joined  \
0  Ольга Петришина                0                 0                0   
1     aadam kathst                0                 0                0   
2    aadish___khan                0                 0                0   
3      Aarzoo Khan                0                 1                1   
4    Abbas Yadavad                0                 0                0   

   total_likes  total_comments  total_posts  follow_back  
0           61               7            2            1  
1           43               0           12            1  
2          184               6           18            0  
3            3               0            1            0  
4            3               0            2            0

In [3]:
#total private account and total business account
print("Total private account are:",insta.private_account.sum())
print("Total business accounts are:",insta.business_account.sum())

Total private account are: 2
Total business accounts are: 11


In [4]:
#count of new users
print("New users:",insta.recently_joined.sum())

New users: 8


In [5]:
#username with maximum followers
print("User with maxmimum followers is {} with total followers:{}".format(
    insta["name"][insta.total_followers==insta.total_followers.max()].values,insta.total_followers.max()))

User with maxmimum followers is ['Sahil Vats'] with total followers:3574


In [6]:
#username with minimum followers
print("User with minimum followers is {} with total followers:{}".format(
    insta["name"][insta.total_followers==insta.total_followers.min()].values,insta.total_followers.min()))


User with minimum followers is ['Shankar Singh Rawat'] with total followers:4


In [7]:
#username with maximum following
print("User with maximum follwing is {} with total following:{}".format(
    insta["name"][insta.total_following==insta.total_following.max()].values,insta.total_following.max()))

User with maximum follwing is ['Prakash'] with total following:3295


In [8]:
#username with minimum following
print("User with minimum following is {} with total following:{}".format(
    insta["name"][insta.total_following==insta.total_following.min()].values,insta.total_following.min()))


User with minimum following is ['Raushan Roy'] with total following:2


In [9]:
#highest number of mutual connections
print("User with highest mutual are {} with {} mutuals".format(
    insta["name"][insta.mutual_connecting==insta.mutual_connecting.max()].values,insta.mutual_connecting.max()))

User with highest mutual are ['ramesh mali 8055@' 'Rahul Upadhayay'] with 11 mutuals


In [10]:
#user with highest likes
print("User with highest likes are {} with {} likes".format(
    insta["name"][insta.total_likes==insta.total_likes.max()].values,insta.total_likes.max()))

User with highest likes are ['Sahil Vats'] with 32252 likes


In [11]:
#user with lowest likes
print("User with lowest likes are {} with {} likes".format(
    insta["name"][insta.total_likes==insta.total_likes.min()].values,insta.total_likes.min()))

User with lowest likes are ['Shankar Singh Rawat'] with 0 likes


In [12]:
#user with maximum comments
print("User with maximimum comments are {} with {} comments".format(
    insta["name"][insta.total_comments==insta.total_comments.max()].values,insta.total_comments.max()))

User with maximimum comments are ['Sahil Vats'] with 2084 comments


In [13]:
#user with minimum comments
print("User with minimum comments are {} with {} comments".format(
    insta["name"][insta.total_comments==insta.total_comments.min()].values,insta.total_comments.min()))

User with minimum comments are ['aadam kathst' 'Aarzoo Khan' 'Abbas Yadavad' 'Abhijit Das'
 'Alexandr Shegolev' 'Arun Siva' 'aryan ranjit' 'Bablu Khatik'
 'balu yadav' 'Ankit _baria' 'Bheru Lal Yadav' 'Bungte Rhiru Roshan'
 'Dev Dadheech' 'Danish Ansari' 'parmar dasharath' 'md_21' 'dav chauhan'
 'emin.mamedov' 'govind' 'Govind Kumar' 'валентина григорук'
 'Hareram kumar singh' 'Harshranjan' 'Ilyas Kineev' 'Irshad Alam'
 'Jitendra Kumar Yadav' 'Kamaraj Poongodi' 'Kamlesh Dadheech' 'Ajay chan'
 'mahendra kathat' 'Kirti Thakor' 'kaushal' 'laishramechan'
 'Людмила Масакова' 'Mahendra Kumae Mahendra Kumar' 'manish saheriya'
 'meera diagnostics' 'Muntun yadav' 'Nelsonam Kujur' 'Nitish YaDav'
 'श्याम लाल नंदराय' 'Pawan' 'pawan Kumar mahto' 'pawansahu' 'Rafik Kathat'
 'Ragho Kumar Ragho Kumar' 'Rajendra Prasad Gupta' 'Ram Pareek'
 'Ramesh Chandra' 'Raushan Roy' 'saran.official' 'sarowarthakur'
 'Senthil Krish' 'Shankar Singh Rawat' 'Sohel Singhaniya'
 'Wahengbam Ton Meitei' 'Tousif Mulla' 'Vig

In [14]:
#user with highest number of posts 
print("User with highest posts are {} with {} posts".format(
    insta["name"][insta.total_posts==insta.total_posts.max()].values,insta.total_posts.max()))

User with highest posts are ['Ramesh Pratap Singh'] with 923 posts


In [15]:
#user with lowest posts
print("User with lowest posts are {} with {} posts".format(
    insta["name"][insta.total_posts==insta.total_posts.min()].values,insta.total_posts.min()))

User with lowest posts are ['Aarzoo Khan' 'Abhijit Das' 'abhishek choudhary' 'Alexandr Shegolev'
 'Amaresh mishra' 'Arun Siva' 'Ashraf khan' 'Bablu Khatik' 'balu yadav'
 'Bungte Rhiru Roshan' 'Deepak joshi...' 'Dev Dadheech' 'Danish Ansari'
 'solanki dev' 'emin.mamedov' 'рафаил' 'govind' 'Govind Kumar'
 'валентина григорук' 'Harshranjan' 'Kamaraj Poongodi' 'Карен'
 'mahendra kathat' 'Kirankiran8872' 'Kirti Thakor' 'kaushal'
 'laishramechan' 'Людмила Масакова' 'Mahendra Kumae Mahendra Kumar'
 'Naveen Nav' 'Nitish YaDav' 'pavanlotti' 'pawan Kumar mahto' 'PMŔv'
 'pragya ajit karn' 'Rafik Kathat' 'Raj Sahu' 'rajsingh' 'Ram Pareek'
 'Ramesh Chandra' 'R K Kumar' 'Raushan Roy' 'R K Kumar' 'saran.official'
 'Satu Mali' 'Shankar Singh Rawat' 'slawa' 'Tanishq Bhai'
 'Wahengbam Ton Meitei' 'Tousif Mulla' 'Вячеслав Валов' 'Vinay Chouhan'
 'Yaseen Mulla' 'Yogesh Nath'] with 1 posts


In [16]:
#total number of followbacks
print(insta["follow_back"].sum())

33


## Model building

In [17]:
insta.columns

Index(['username', 'mutual_connecting', 'total_following', 'total_followers',
       'name', 'private_account', 'business_account', 'recently_joined',
       'total_likes', 'total_comments', 'total_posts', 'follow_back'],
      dtype='object')

In [65]:
#storing target variable in a different variable
x = insta.drop(["username","name","follow_back"],axis=1)
y = insta["follow_back"]

In [68]:
#splitting our data into train and test
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size = 0.3,random_state = 123)

## Using Random Forest Classifier to build a base model

In [153]:
#importing and building a base random forest model
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier(random_state=10)
rf.fit(x_train,y_train)

c:\ProgramData\Anaconda3\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=None,
            oob_score=False, random_state=10, verbose=0, warm_start=False)

In [154]:
#checking metrics
from sklearn import metrics
metrics.accuracy_score(y_test,rf.predict(x_test))

0.8214285714285714

In [155]:
metrics.roc_auc_score(y_test,rf.predict(x_test))

0.5833333333333333

## Hyperparameter tuning

In [77]:
from sklearn.model_selection import GridSearchCV

In [173]:
#defining parameter for using in grid search cv
param_grids = {"n_estimators":np.arange(10,100,10),
              "min_samples_leaf":np.arange(1,50,1),
              "oob_score":[True,False]}

In [174]:
rf_grid = GridSearchCV(RandomForestClassifier(), param_grids, cv = 5, scoring="roc_auc", n_jobs=-1)

In [175]:
rf_grid.fit(x_train,y_train)

GridSearchCV(cv=5, error_score='raise-deprecating',
       estimator=RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators='warn', n_jobs=None,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False),
       fit_params=None, iid='warn', n_jobs=-1,
       param_grid={'n_estimators': array([10, 20, 30, 40, 50, 60, 70, 80, 90]), 'min_samples_leaf': array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49]), 'oob_score': [True, False]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring='roc_auc', verbose=0)

In [176]:
rf_grid.best_params_

{'min_samples_leaf': 31, 'n_estimators': 30, 'oob_score': True}

In [177]:
#roc auc score
metrics.roc_auc_score(y_test,rf_grid.predict(x_test))

0.5

In [178]:
#accuracy
metrics.accuracy_score(y_test,rf_grid.predict(x_test))

0.8571428571428571

## Building final model using best parameters

In [179]:
model = RandomForestClassifier(min_samples_leaf = 33 ,n_estimators = 90,oob_score=False)
model.fit(x_train,y_train)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=33, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=90, n_jobs=None,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False)

In [180]:
y_predicted = model.predict(x_test)

In [181]:
metrics.accuracy_score(y_test,y_predicted)

0.8571428571428571

In [182]:
metrics.roc_auc_score(y_test,y_predicted)

0.5

### Our model has an accuracy score of 85.7% 

# END